### PR TITLE
setops v3

### PR DESCRIPTION
**Pull request by Nekomajin42**
Implement (multi)set operations with two helper functions + tests

### CODE REVIEW

**Comment by Nekomajin42** [comment]
Hi Jos!

Sorry for the delay and the new pr, but I had some difficulties on my PC.

I hope I've done everything we discussed in #756.

Apart from the new functions, there are two helper functions in `utils/array.js`. I've added tests for the matrix output and also for the helper functions. I hope it will work now as expected.

**Comment by josdejong** [comment]
Thanks Toth! I will review your PR asap but please give me some time, its a lot of code :)

File: lib/function/set/setDistinct.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    'Array | Matrix': function (a) {
+      if (subset(size(a), new index(0)) === 0) { // if empty, return empty
+        var result = [];
+      }
+      else {
+        var b = Array.isArray(a) ? flatten(a) : flatten(a.toArray());
+        var result = [];
+        result.push(b[0]);
+        for (var i=1; i<b.length; i++) {
+          if (!equal(b[i], b[i-1])) {
```
Comment (samczsun) [inline lib/function/set/setDistinct.js]: This doesn't seem like it would work with an input like `[1,2,1]`

File: lib/function/set/setSize.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    },
+    'Array | Matrix, boolean': function (a, unique) {
+      if (unique === false || a.length === 0) {
+        return Array.isArray(a) ? flatten(a).length : flatten(a.toArray()).length;
+      }
+      else {
+        var b = Array.isArray(a) ? flatten(a) : flatten(a.toArray());
+        var count = 1;
+        for (var i=1; i<b.length; i++) {
+          if (!equal(b[i], b[i-1])) {
```
Comment (samczsun) [inline lib/function/set/setSize.js]: Same issue as `setDistinct`

File: lib/utils/array.js [hunk 3]
```diff
@@ -446,6 +446,40 @@ exports.join = function (array, separator) {
...
 
 /**
+ * Assign a numeric identifier to every element of an array
+ * @param {Array}	a  An array
+ * @return {Array}	An array of objects containing the original value and its identifier
+ */
+exports.identify = function(a) {
+  var b = [];
+  var count = 0;
+  b[0] = {value: a[0], identifier: 0};
```
Comment (samczsun) [inline lib/utils/array.js]: What if `a` is not an array or `a.length == 0`?

File: lib/utils/array.js [hunk 4]
```diff
@@ -446,6 +446,40 @@ exports.join = function (array, separator) {
...
+  for (var i=1; i<a.length; i++) {
+    if (a[i] === a[i-1]) {
```
Comment (samczsun) [inline lib/utils/array.js]: This does not seem to produce the correct result for `[1, 2, 1]`

**Comment by Nekomajin42** [comment]
@samczsun
Thanks for pointing these out! At first, I sorted the arrays before doing anything with them, but because of the issue regarding complex numbers, I removed the sorting step. With sorted arrays, all the functions should work as intended.

@josdejong 
What is the decision with the sorting of complex numbers? How do you thing we should handle this?

**Comment by josdejong** [comment]
@Nekomajin42 we decided to postpone implementing comparing complex numbers until there is a concrete use case for it (see #837). From our discussion https://github.com/josdejong/mathjs/pull/756#issuecomment-277294507 I understood that you can adjust the functions such that they don't need ordering?

**Comment by Nekomajin42** [comment]
@josdejong 
Well, the problem comes when the input contains a multiset, when I have to identify each element, so I can treat the same values as unique elements. That's why I had to implement the helper functions. I think, I can modify the `identify` function to work with unsorted arrays by comparing every element with the previous ones, but I don't know how it would effect the efficiency.

In my use case, the ascending or descending order is not important, only to have the same values next to each other, so I can append a numeric index to each.

What do you think about treating the complex numbers as 2D coordinates of a vector and sorting them by their absolute length? Or first sorting by the real part, and then sorting the same values by the imaginary part again? If I understand from the discussion you've mentioned, there is no mathematically correct way of sorting complex numbers. But as a documented workaround, I think it would work well.

**Comment by josdejong** [comment]
Ok so we _do_ need ordering of complex numbers else the `identity` function becomes poorly performing. 

Both options for ordering complex numbers (by real value or by absolute length) do not guarantee a unique ordering. For example when ordering by real value, `[1+i, 2+i, 3+3i]` and `[1+i, 3+3i, 2+i]` contain the same values and are both ordered, but not the same. Is using such an ordering of complex numbers an option anyway?

**Comment by Nekomajin42** [comment]
How is the second array ordered?

I've meant something like this: `[-1+i, -1+2i, 1+i, 1+2i, 2-2i, 2-i, ...]`. Just a simple two level sorting. First by the real part, and if two has the same real part, then sort them again by the imaginary part.

As I think more about it, the absolute length would not work because `1+i` and `1-i` would be the same. But the other solution should work.

**Comment by josdejong** [comment]
Ah, sorry, I messed up the example of the two arrays. I've updated them to have two values both having a real part 2 and a different imaginary part.

> As I think more about it, the absolute length would not work because 1+i and 1-i would be the same. 

Yes that's what I meant. That's a problem with both sorting by absolute length or by real value: it will not uniquely sort differing values like `1+i` and `1-i`. 

Your idea of sorting by real part first and imaginary part second sounds like a great idea, that will solve this issue! And I think it will look more logical to the human eye.

I will reopen #837 then

**Comment by Nekomajin42** [comment]
OK! I will wait for you to implement the sorting, then modify my code to sort before identifying.

**Comment by josdejong** [comment]
@Nekomajin42 , @gulfaraz has implemented comparison of complex numbers. This is now available in the `develop` branch, so I think you can go ahead and use it in the set functions.

**Comment by josdejong** [comment]
@Nekomajin42 I see this PR is still open, but it's already redundant and can be closed right?

**Pull request closed without merging**